### PR TITLE
LAM-2222: added validation for preview route placeholder

### PR DESCRIPTION
**Pull request by sandroJayas**

### CODE REVIEW

File: cmd/adapters/ingresses/ingresses.go [hunk 1]
```diff
@@ -107,6 +116,32 @@ func (c *VampCloudAnansiIngressClient) PostIngress(ingress models.Ingress) (int6
...
+func checkPreviewRoute(routes []*dto.Route) error {
+
+	re := regexp.MustCompile(`%%(.*)%%`)
+	for _, v := range routes {
+		path := strings.Split(v.Path, "/")
+
+		for _, v2 := range path {
+			result := re.FindSubmatch([]byte(v2))
+			if len(result) > 0 && !validPlaceholder(result[1]) {
+				return fmt.Errorf("placeholder validation failed")
```
Comment (dmitry-onishchenko) [inline cmd/adapters/ingresses/ingresses.go:128]: the minor comment - does it make sense to make the error message more obvious, e.g. "placeholder '%s' is not supported"

**Review by dmitry-onishchenko — CHANGES_REQUESTED** [review]
I think we need cover the functionality with tests

File: cmd/adapters/ingresses/ingresses.go [hunk 2]
```diff
@@ -115,7 +150,11 @@ func (c *VampCloudAnansiIngressClient) PatchIngress(ingress models.Ingress) erro
 
 	params := operations.NewPatchApplicationsApplicationIDIngressesIngressIDParams().WithApplicationID(ingress.ApplicationID).WithIngressID(ingress.ID).WithIngress(&ingressInput)
 
-	_, err := c.client.Operations.PatchApplicationsApplicationIDIngressesIngressID(params, nil)
+	err := checkPreviewRoute(params.Ingress.Routes)
+	if err != nil {
+		return fmt.Errorf("failed to create ingress: %w", err)
```
Comment (dmitry-onishchenko) [inline cmd/adapters/ingresses/ingresses.go:155]: I think it should be "failed to patch ingress"

File: cmd/adapters/ingresses/ingresses.go [hunk 3]
```diff
@@ -107,6 +116,32 @@ func (c *VampCloudAnansiIngressClient) PostIngress(ingress models.Ingress) (int6
...
 
+func checkPreviewRoute(routes []*dto.Route) error {
+
+	re := regexp.MustCompile(`%%(.*)%%`)
+	for _, v := range routes {
+		path := strings.Split(v.Path, "/")
+
+		for _, v2 := range path {
+			result := re.FindSubmatch([]byte(v2))
+			if len(result) > 0 && !validPlaceholder(result[1]) {
```
Comment (avalcepina) [inline cmd/adapters/ingresses/ingresses.go:127]: You can have more than a single occurrence of a placeholder

Comment (dmitry-onishchenko) [inline cmd/adapters/ingresses/ingresses.go:127]: * shall we check length against 1 instead, e.g. `len(result) > 1` just to make sure that result[1] is always safe?
* why not to use `FindStringSubmatch ` instead? (to skip type casting)

File: cmd/adapters/ingresses/ingresses.go [hunk 4]
```diff
@@ -107,6 +116,32 @@ func (c *VampCloudAnansiIngressClient) PostIngress(ingress models.Ingress) (int6
 
 }
 
+func checkPreviewRoute(routes []*dto.Route) error {
+
+	re := regexp.MustCompile(`%%(.*)%%`)
+	for _, v := range routes {
+		path := strings.Split(v.Path, "/")
```
[the remark this answers is not in the document]
Comment (sandroJayas) [reply cmd/adapters/ingresses/ingresses.go:123]: `some-%%path/%%VERSION%%` is a valid path, without split it should be considered invalid

File: cmd/adapters/ingresses/ingresses.go (hunk 3 above)
Comment (sandroJayas) [reply cmd/adapters/ingresses/ingresses.go:127]: how?

**Review by dmitry-onishchenko — APPROVED** [review]
LGTM

**Review by avalcepina — APPROVED** [review]

**Pull request merged**